### PR TITLE
1888-opacLink

### PR DESCRIPTION
**Pull request by TobiasNx**
This PR does the following:
- Refactoring the mapping `heldBy` so that the ISIL of the IZ is provided, that changes the order of `heldBy.isil` and `heldBy.id`
- Adding a macro for adding a `opacLink` as `sameAs`
- Adding mapping files (tsv) for `opacLink` tusing a submodule for prod and a local copy for test
- Adding `sameAs` for all mapped `hasItem`-objects

@acka47 do you have a better property than `opacLink`?

-----
Outdated:
~First draft of opac link enrichment mechanism~

~@blackwinter: my first design runs into an error, could you have a look at it:~

<details>
<summary>~Error-message~</summary>
```
[main] ERROR org.lobid.resources.AlmaMarc21XmlToLobidJsonMetafixTest - Errored when transforming 
org.metafacture.metafix.FixProcessException: Error while executing Fix expression (at file:/home/tobias/git/lobid-resources/src/main/resources/alma/fix/macros.fix, line 267): split_field("$i.@opacLink","{hbzid}")
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:249)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:125)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.lambda$null$14(RecordTransformer.java:183)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:226)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:125)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:237)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:125)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:116)
        at org.metafacture.metafix.FixMethod$10.apply(FixMethod.java:155)
        at org.metafacture.metafix.RecordTransformer.lambda$null$20(RecordTransformer.java:206)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:226)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:125)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:237)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:125)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.lambda$null$17(RecordTransformer.java:197)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:226)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:125)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:237)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:125)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:124)
        at org.metafacture.metafix.FixBind$1.lambda$execute$1(FixBind.java:40)
        at org.metafacture.metafix.Value.asList(Value.java:190)
        at org.metafacture.metafix.Value.asList(Value.java:184)
        at org.metafacture.metafix.FixBind$1.execute(FixBind.java:33)
        at org.metafacture.metafix.RecordTransformer.lambda$null$4(RecordTransformer.java:138)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:226)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:125)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:237)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:125)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:124)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:116)
        at org.metafacture.metafix.FixMethod$1.apply(FixMethod.java:60)
        at org.metafacture.metafix.RecordTransformer.lambda$null$20(RecordTransformer.java:206)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:226)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:125)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:237)
        at org.metafacture.metafix.RecordTransformer.lambda$transform$2(RecordTransformer.java:125)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:124)
        at org.metafacture.metafix.Metafix.endRecord(Metafix.java:222)
        at org.metafacture.monitoring.StreamBatchLogger.endRecord(StreamBatchLogger.java:129)
        at org.metafacture.biblio.marc21.MarcXmlHandler.endElement(MarcXmlHandler.java:132)
        at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:609)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1781)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2966)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:601)
        at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:504)
        at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:841)
        at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:770)
        at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
        at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:69)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:43)
        at org.metafacture.io.FileOpener.process(FileOpener.java:158)
        at org.metafacture.io.FileOpener.process(FileOpener.java:41)
        at org.metafacture.files.DirReader.dir(DirReader.java:99)
        at org.metafacture.files.DirReader.process(DirReader.java:75)
        at org.lobid.resources.AlmaMarc21XmlToLobidJsonMetafixTest.transformFile(AlmaMarc21XmlToLobidJsonMetafixTest.java:124)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: java.util.regex.PatternSyntaxException: Illegal repetition
{hbzid}
        at java.util.regex.Pattern.error(Pattern.java:1969)
        at java.util.regex.Pattern.closure(Pattern.java:3171)
        at java.util.regex.Pattern.sequence(Pattern.java:2148)
        at java.util.regex.Pattern.expr(Pattern.java:2010)
        at java.util.regex.Pattern.compile(Pattern.java:1702)
        at java.util.regex.Pattern.<init>(Pattern.java:1352)
        at java.util.regex.Pattern.compile(Pattern.java:1028)
        at org.metafacture.metafix.FixMethod$43.apply(FixMethod.java:595)
        at org.metafacture.metafix.RecordTransformer.lambda$null$20(RecordTransformer.java:206)
        at org.metafacture.metafix.RecordTransformer.lambda$null$22(RecordTransformer.java:226)
        at org.metafacture.metafix.RecordTransformer.lambda$null$1(RecordTransformer.java:125)
        at org.metafacture.metafix.RecordTransformer.tryRun(RecordTransformer.java:237)
        ... 103 more
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.088 sec <<< FAILURE! - in UnitTests
```
</details>

### CODE REVIEW

**Comment by blackwinter** [comment]
> Caused by: java.util.regex.PatternSyntaxException: Illegal repetition

The pattern in `split_field` is interpreted as a regular expression. You need to escape the curly braces.

**Comment by TobiasNx** [comment]
@blackwinter thanks.
There seems to be a problem with `copy_field` (but also `paste`) a copied and manipulated field (here @iz) within a list bind.
The copy of the copied field is not manipulated (not turned into a ISIL), therefore the new lookup for the opac link does not work.

This seems to be a fix issue. The last commit highlights the problem.

File: src/test/resources/alma-fix/990156060190206441.json [hunk 1]
```diff
@@ -222,6 +226,7 @@
     "serialNumber" : "91 AVY 2",
     "currentLibrary" : "EUP",
     "currentLocation" : "EUP_BV",
+    "opacLink" : "https://onlinekatalog.bibliothek.fh-aachen.de/webOPACClient/start.do?Query=-1=%22%22",
```
Comment (TobiasNx) [inline src/test/resources/alma-fix/990156060190206441.json:229]: @blackwinter this is an example of the failing paste. I am not sure why.

Comment (blackwinter) [reply src/test/resources/alma-fix/990156060190206441.json:229]: The title doesn't have a `zdbId`. You should wrap your `opacLink` lookups in `exists("xxxId")` (and preferably reverse the order, i.e. `hbzId` before `zdbId`).

Comment (TobiasNx) [reply src/test/resources/alma-fix/990156060190206441.json]: Thanks a lot! This worked!

Comment (blackwinter) [reply src/test/resources/alma-fix/990156060190206441.json]: Please note that you also have to check if `hbzId` exists. Not all records have one.

Comment (TobiasNx) [reply src/test/resources/alma-fix/990156060190206441.json]: You are right, we also need to exclude hbzId when zdbId exists.

File: src/main/resources/alma/fix/macros.fix [hunk 2]
```diff
@@ -259,14 +259,29 @@ do put_macro("opacLink")
...
   end
-  copy_field("$i.@iz", "$i.@opacLinkHbzId")
-  lookup("$i.@opacLinkHbzId","isil2opac_hbzId", delete:"true")
-  split_field("$i.@opacLinkHbzId","\\{hbzid\\}")
   if exists("$i.@opacLinkZdbId.1")
     paste("hasItem[].$last.opacLink","$i.@opacLinkZdbId.1","zdbId","$i.@opacLinkZdbId.2",join_char:"")
   elsif exists("$i.@opacLinkHbzId.1")
     paste("hasItem[].$last.opacLink","$i.@opacLinkHbzId.1","hbzId","$i.@opacLinkHbzId.2",join_char:"")
+  elsif exists("$i.@opacLinkIsbn.1")
+    paste("hasItem[].$last.opacLink","$i.@opacLinkIsbn.1","isbn[].1","$i.@opacLinkisbn.2",join_char:"")
```
Comment (blackwinter) [inline src/main/resources/alma/fix/macros.fix:282]:
```suggestion
    paste("hasItem[].$last.opacLink","$i.@opacLinkIsbn.1","isbn[].1","$i.@opacLinkIsbn.2",join_char:"")
```

File: src/main/resources/alma/fix/macros.fix [hunk 3]
```diff
@@ -259,14 +259,29 @@ do put_macro("opacLink")
...
+  elsif exists("$i.@opacLinkIssn.1")
+    paste("hasItem[].$last.opacLink","$i.@opacLinkIssn.1","issn[].1","$i.@opacLinkissn.2",join_char:"")
```
Comment (blackwinter) [inline src/main/resources/alma/fix/macros.fix:284]:
```suggestion
    paste("hasItem[].$last.opacLink","$i.@opacLinkIssn.1","issn[].1","$i.@opacLinkIssn.2",join_char:"")
```

**Review by blackwinter — CHANGES_REQUESTED** [review]

**Comment by TobiasNx** [comment]
https://primo.kobv.de/primo_library/libweb/action/dlSearch.do?vid=TRI_UB_I&institution=TRI_UB_I&search_scope=localude&bulkSize=10&lang=ger&indx=1&onCampus=false&query=any,contains,HT008733617

These links seem to be outdated.

**Comment by blackwinter** [comment]
> These links seem to be outdated.

Related to hbz/link-templates#21.

**Comment by blackwinter** [comment]
> Adding `opacUrl` for all mapped `hasItem`-objects

The newly introduced property is currently named `opacLink`.

**Review by blackwinter — APPROVED** [review]
Looks good, thanks!

File: src/main/resources/alma/fix/macros.fix [hunk 4]
```diff
@@ -252,6 +252,40 @@ do put_macro("provenanceLinks")
...
+    copy_field("$i.$[field]", "$i.@opacLinkIssn")
+    lookup("$i.@opacLinkIssn","isil2opac_issn", delete:"true")
+    split_field("$i.@opacLinkIssn","\\{issn\\}")
+  end
+  if exists("isbn[].1")
+    copy_field("$i.$[field]", "$i.@opacLinkIsbn")
+    lookup("$i.@opacLinkIsbn","isil2opac_isbn", delete:"true")
+    split_field("$i.@opacLinkZdbId","\\{isbn\\}")
+  end
+  set_array("hasItem[].$last.seeAlso[]")
```
Comment (blackwinter) [inline src/main/resources/alma/fix/macros.fix:277]: Does this mean that there can be multiple links in the future?

File: src/main/resources/alma/fix/item.fix [hunk 5]
```diff
@@ -18,26 +18,28 @@ do list(path:"ITM  ", "var": "$i")
...
-      copy_field("$i.a", "hasItem[].$last.heldBy.id")
-      replace_all("hasItem[].$last.heldBy.id",".*(\\d{4})$","$1")
-      lookup("hasItem[].$last.heldBy.id", "alma-institution-code-to-isil")
+      copy_field("$i.@iz","hasItem[].$last.heldBy.isil")
     end
-    copy_field("hasItem[].$last.heldBy.id","hasItem[].$last.heldBy.isil")
-    prepend("hasItem[].$last.heldBy.id", "http://lobid.org/organisations/")
-    append("hasItem[].$last.heldBy.id","#!")
+    call_macro("opacLink", field:"@sublibraryIsil")
+    unless exists("hasItem[].$last.opacLink")
```
Comment (blackwinter) [inline src/main/resources/alma/fix/item.fix:39]:
```suggestion
    unless exists("hasItem[].$last.seeAlso[]")
```

File: src/main/resources/alma/fix/item.fix [hunk 6]
```diff
@@ -59,26 +61,28 @@ do list(path: "HOL  ", "var": "$i")
...
+            copy_field("$i.@sublibraryIsil", "hasItem[].$last.heldBy.isil")
           # if no mapping for a sublocation code is provided or no sublocation code exists ($i.w) the main library ISIL is used.
           else
-            copy_field("$i.8", "hasItem[].$last.heldBy.id")
-            replace_all("hasItem[].$last.heldBy.id",".*(\\d{4})$","$1")
-            lookup("hasItem[].$last.heldBy.id", "alma-institution-code-to-isil")
+            copy_field("$i.@iz", "hasItem[].$last.heldBy.isil")
+          end
+          call_macro("opacLink", field:"@sublibraryIsil")
+          unless exists("hasItem[].$last.opacLink")
```
Comment (blackwinter) [inline src/main/resources/alma/fix/item.fix:82]:
```suggestion
          unless exists("hasItem[].$last.seeAlso[]")
```

**Review by blackwinter — CHANGES_REQUESTED** [review]

File: src/main/resources/alma/fix/macros.fix (hunk 4 above)
Comment (TobiasNx) [reply src/main/resources/alma/fix/macros.fix:277]: nope, but the property is defined as set in the context.

Comment (blackwinter) [reply src/main/resources/alma/fix/macros.fix:277]: Okay, I'll take your word for it ;)

**Comment by TobiasNx** [comment]
There is one last change @blackwinter

File: src/main/resources/alma/fix/macros.fix (hunk 4 above)
Comment (TobiasNx) [reply src/main/resources/alma/fix/macros.fix:280]: Nope, the context says it has to be an array, since it is used elsewhere in the record too.

Comment (blackwinter) [reply src/main/resources/alma/fix/macros.fix:280]: I understand that. I meant that I trust that you don't sneak additional URLs in there in the future.

File: src/main/resources/alma/fix/macros.fix [hunk 7]
```diff
@@ -279,6 +282,8 @@ do put_macro("opacLink")
     paste("hasItem[].$last.seeAlso[].$append","$i.@opacLinkZdbId.1","zdbId","$i.@opacLinkZdbId.2",join_char:"")
   elsif exists("$i.@opacLinkHbzId.1")
     paste("hasItem[].$last.seeAlso[].$append","$i.@opacLinkHbzId.1","hbzId","$i.@opacLinkHbzId.2",join_char:"")
+  elsif exists("$i.@opacAlmaMmsId.1")
+    paste("hasItem[].$last.seeAlso[].$append","$i.@opacAlmaMmsId.1","hbzId","$i.@opacAlmaMmsId.2",join_char:"")
```
Comment (blackwinter) [inline src/main/resources/alma/fix/macros.fix:286]:
```suggestion
    paste("hasItem[].$last.seeAlso[].$append","$i.@opacAlmaMmsId.1","almaMmsId","$i.@opacAlmaMmsId.2",join_char:"")
```

**Review by dr0i — CHANGES_REQUESTED** [review]
putting context.jsonld is not necessary I think? Please remove it. It brings conflicts, as discussed . There is hope so - the ordering of the context!

**Comment by blackwinter** [comment]
I'm not exactly sure what conflicts you're talking about, but does 7c6fb57 help?

**Comment by dr0i** [comment]
thx @blackwinter!

**Comment by dr0i** [comment]
Will be deployed next Monday.

**Comment by blackwinter** [comment]
This didn't work as expected. E.g. [990366624790206441](https://lobid.org/resources/990366624790206441):

```json
"hasItem" : [ {
  // ...
  "seeAlso" : [ "DE-Hag4HT020469610" ],
  "id" : "http://lobid.org/items/990366624790206441:DE-Hag4:5378132900006461#!"
}, {
  // ...
  "seeAlso" : [ "DE-575HT020469610" ],
  "id" : "http://lobid.org/items/990366624790206441:DE-575:53102917670006480#!"
} ],
```

I couldn't find any link that was added correctly.

**Comment by dr0i** [comment]
Right, that was because gitsubmodules (since https://github.com/hbz/lobid-resources/pull/1902  ) didnt't update the isil2opac (https://github.com/hbz/lookup-tables/commit/19c5676bae0c61b2a08716deed36d904df5496a4). I've restarted the indexing, should be deployed tomorrow.

**Comment by blackwinter** [comment]
Caused by missing file map: `lookup()` calls `map.get()` which tries `FileMap.init()` leading to an exception _every_ time. This then leaves the looked up value untouched even though `delete: "true"` was specified.

**Pull request merged**